File: app/routers/mf_breakdown.py
import asyncio

from fastapi import APIRouter, Depends, Request
from fastapi.responses import JSONResponse
from sqlalchemy import update
from sqlalchemy.dialects.postgresql import insert as pg_insert
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.sse import sse_stream
from app.models.allocation_target import AllocationTarget
from app.models.mf_breakdown import EquityCategoryOverride, MfSchemeBreakdown
from app.services.mf_breakdown import (
    normalize_company_name,
    get_allocation_comparison,
    get_allocation_targets,
    get_asset_class_comparison,
    get_asset_class_targets,
    get_available_schemes,
    get_breakdown_chart_data,
    get_category_composition,
    get_direct_trade_breakdown,
    get_rebalance_plan,
    get_scheme_breakdown,
    get_sector_composition,
    get_sector_list,
    get_sector_stock_breakdown,
    get_stock_holdings_table,
    ingest_from_openfin,
    save_allocation_targets,
    save_asset_class_targets,
    save_sector_overrides,
    sync_amfi_market_cap,
)
from app.time_util import now_ist

router = APIRouter(prefix="/api/v1/mf-breakdown", tags=["mf-breakdown"])

_ingest_lock = asyncio.Lock()


@router.get("/ingest/stream")
async def ingest_stream(db: AsyncSession = Depends(get_db)):
    async def _runner(on_progress):
        amfi = await sync_amfi_market_cap(db, on_progress=on_progress)
        if "error" not in amfi:
            ingest = await ingest_from_openfin(db, on_progress=on_progress)
        else:
            ingest = {"error": "Skipped — AMFI classification not loaded"}
        return {"amfi": amfi, "ingest": ingest}

    return sse_stream(_runner, lock=_ingest_lock, busy_msg="An ingest is already running.")


VALID_CATEGORIES = {"Large Cap", "Mid Cap", "Small Cap", "Unclassified Equity", "Equity - Foreign"}


@router.patch("/classify-batch")
async def classify_batch(request: Request, db: AsyncSession = Depends(get_db)):
    form = await request.form()
    isins = form.getlist("scheme_isin")
    names = form.getlist("name")
    categories = form.getlist("category")

    updated = 0
    override_rows: list[dict] = []
    for isin, name, cat in zip(isins, names, categories):
        if cat not in VALID_CATEGORIES:
            continue
        result = await db.execute(
            update(MfSchemeBreakdown)
            .where(MfSchemeBreakdown.scheme_isin == isin, MfSchemeBreakdown.name == name)
            .values(category=cat, updated_at=now_ist())
        )
        updated += result.rowcount
        override_rows.append({
            "name_normalized": normalize_company_name(name),
            "raw_name": name,
            "category": cat,
            "updated_at": now_ist(),
        })

    if override_rows:
        stmt = pg_insert(EquityCategoryOverride).values(override_rows)
        await db.execute(stmt.on_conflict_do_update(
            index_elements=["name_normalized"],
            set_={"raw_name": stmt.excluded.raw_name, "category": stmt.excluded.category, "updated_at": stmt.excluded.updated_at},
        ))
    await db.commit()

    return {"updated": updated}


@router.patch("/sector-classify-batch")
async def sector_classify_batch(request: Request, db: AsyncSession = Depends(get_db)):
    body = await request.json()
    updated = await save_sector_overrides(db, body)
    return {"updated": updated}


@router.get("/sector-list")
async def sector_list(db: AsyncSession = Depends(get_db)):
    sectors = await get_sector_list(db)
    return JSONResponse(sectors)


@router.get("/chart-data")
async def chart_data(db: AsyncSession = Depends(get_db)):
    data = await get_breakdown_chart_data(db)
    return JSONResponse(data)


@router.get("/stock-holdings")
async def stock_holdings(db: AsyncSession = Depends(get_db)):
    data = await get_stock_holdings_table(db)
    return JSONResponse(data)


@router.get("/allocation-comparison")
async def allocation_comparison(mode: str = "anchored", db: AsyncSession = Depends(get_db)):
    data = await get_allocation_comparison(db, mode=mode)
    return JSONResponse(data)


@router.get("/allocation-targets")
async def allocation_targets(mode: str = "anchored", db: AsyncSession = Depends(get_db)):
    targets = await get_allocation_targets(db, mode=mode)
    return JSONResponse(targets)


@router.post("/allocation-targets")
async def update_allocation_targets(request: Request, db: AsyncSession = Depends(get_db)):
    form = await request.form()
    mode = form.get("mode", "anchored")
    targets: dict[str, float] = {}
    for key, val in form.items():
        if key.startswith("target_"):
            cat = key[7:]
            try:
                targets[cat] = float(val)
            except ValueError:
                continue
    await save_allocation_targets(db, targets, mode=str(mode))
    return {"ok": True}


@router.get("/rebalance-plan")
async def rebalance_plan(
    mode: str = "anchored",
    cash: float | None = None,
    db: AsyncSession = Depends(get_db),
):
    data = await get_rebalance_plan(db, mode=mode, cash_amount=cash)
    return JSONResponse(data)


@router.get("/asset-class-comparison")
async def asset_class_comparison(db: AsyncSession = Depends(get_db)):
    data = await get_asset_class_comparison(db)
    return JSONResponse(data)


@router.get("/asset-class-targets")
async def get_asset_class_targets_endpoint(db: AsyncSession = Depends(get_db)):
    targets = await get_asset_class_targets(db)
    return JSONResponse(targets)


@router.post("/asset-class-targets")
async def update_asset_class_targets(request: Request, db: AsyncSession = Depends(get_db)):
    form = await request.form()
    targets: dict[str, float] = {}
    foreign_pct: float | None = None
    for key, val in form.items():
        if key.startswith("target_"):
            asset_class = key[7:].replace("_", " ")
            try:
                v = float(val)
            except ValueError:
                continue
            if asset_class == "Equity - Foreign":
                foreign_pct = v
            else:
                targets[asset_class] = v
    await save_asset_class_targets(db, targets)
    if foreign_pct is not None:
        await db.execute(
            pg_insert(AllocationTarget)
            .values(category="Equity - Foreign", target_pct=foreign_pct)
            .on_conflict_do_update(index_elements=["category"], set_={"target_pct": foreign_pct})
        )
        await db.commit()
    return {"ok": True}


@router.get("/category-composition")
async def category_composition(db: AsyncSession = Depends(get_db)):
    data = await get_category_composition(db)
    return JSONResponse(data)


@router.get("/sector-composition")
async def sector_composition(db: AsyncSession = Depends(get_db)):
    data = await get_sector_composition(db, equity_only=True)
    return JSONResponse(data)


@router.get("/sector-stock-breakdown")
async def sector_stock_breakdown(db: AsyncSession = Depends(get_db)):
    data = await get_sector_stock_breakdown(db)
    return JSONResponse(data)


@router.get("/direct-trades")
async def direct_trades(db: AsyncSession = Depends(get_db)):
    data = await get_direct_trade_breakdown(db)
    return JSONResponse(data)


@router.get("/schemes")
async def schemes(db: AsyncSession = Depends(get_db)):
    data = await get_available_schemes(db)
    return JSONResponse(data)


@router.get("/scheme/{scheme_isin}")
async def scheme_detail(scheme_isin: str, db: AsyncSession = Depends(get_db)):
    data = await get_scheme_breakdown(db, scheme_isin)
    return JSONResponse(data)

File: frontend/src/api/mfBreakdown.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { request, requestForm } from './client'
import type {
  AllocationComparison,
  AssetClassComparison,
  BreakdownChartData,
  CategoryCompositionItem,
  ClassifyResult,
  DirectTradeBreakdown,
  SchemeBreakdown,
  SchemeListItem,
  SectorClassifyResult,
  SectorCompositionItem,
  SectorStockBreakdownItem,
  StockHolding,
} from '../types/mfBreakdown'

export const breakdownKeys = {
  chartData: ['mf-breakdown', 'chart-data'] as const,
  stockHoldings: ['mf-breakdown', 'stock-holdings'] as const,
  allocationComparison: ['mf-breakdown', 'allocation-comparison'] as const,
  allocationTargets: ['mf-breakdown', 'allocation-targets'] as const,
  assetClassComparison: ['mf-breakdown', 'asset-class-comparison'] as const,
  categoryComposition: ['mf-breakdown', 'category-composition'] as const,
  sectorComposition: ['mf-breakdown', 'sector-composition'] as const,
  sectorStockBreakdown: ['mf-breakdown', 'sector-stock-breakdown'] as const,
  sectorList: ['mf-breakdown', 'sector-list'] as const,
  directTrades: ['mf-breakdown', 'direct-trades'] as const,
  schemes: ['mf-breakdown', 'schemes'] as const,
  scheme: (isin: string) => ['mf-breakdown', 'scheme', isin] as const,
}

export function useBreakdownChart() {
  return useQuery({
    queryKey: breakdownKeys.chartData,
    queryFn: () => request<BreakdownChartData>('/api/v1/mf-breakdown/chart-data'),
  })
}

export function useStockHoldings() {
  return useQuery({
    queryKey: breakdownKeys.stockHoldings,
    queryFn: () => request<StockHolding[]>('/api/v1/mf-breakdown/stock-holdings'),
  })
}

export function useAllocationComparison(mode: 'anchored' | 'free_float' = 'anchored') {
  return useQuery({
    queryKey: [...breakdownKeys.allocationComparison, mode],
    queryFn: () => request<AllocationComparison>(`/api/v1/mf-breakdown/allocation-comparison?mode=${mode}`),
  })
}

export function useAllocationTargets() {
  return useQuery({
    queryKey: breakdownKeys.allocationTargets,
    queryFn: () => request<Record<string, number>>('/api/v1/mf-breakdown/allocation-targets'),
  })
}

export function useCategoryComposition() {
  return useQuery({
    queryKey: breakdownKeys.categoryComposition,
    queryFn: () => request<CategoryCompositionItem[]>('/api/v1/mf-breakdown/category-composition'),
  })
}

export function useSectorComposition() {
  return useQuery({
    queryKey: breakdownKeys.sectorComposition,
    queryFn: () => request<SectorCompositionItem[]>('/api/v1/mf-breakdown/sector-composition'),
  })
}

export function useSectorStockBreakdown() {
  return useQuery({
    queryKey: breakdownKeys.sectorStockBreakdown,
    queryFn: () => request<SectorStockBreakdownItem[]>('/api/v1/mf-breakdown/sector-stock-breakdown'),
  })
}

export function useDirectTrades() {
  return useQuery({
    queryKey: breakdownKeys.directTrades,
    queryFn: () => request<DirectTradeBreakdown[]>('/api/v1/mf-breakdown/direct-trades'),
  })
}

export function useAvailableSchemes() {
  return useQuery({
    queryKey: breakdownKeys.schemes,
    queryFn: () => request<SchemeListItem[]>('/api/v1/mf-breakdown/schemes'),
  })
}

export function useSchemeBreakdown(schemeIsin: string | null) {
  return useQuery({
    queryKey: breakdownKeys.scheme(schemeIsin ?? ''),
    queryFn: () => request<SchemeBreakdown>(`/api/v1/mf-breakdown/scheme/${schemeIsin}`),
    enabled: schemeIsin != null,
  })
}

export function useAssetClassComparison() {
  return useQuery({
    queryKey: breakdownKeys.assetClassComparison,
    queryFn: () => request<AssetClassComparison>('/api/v1/mf-breakdown/asset-class-comparison'),
  })
}

export function useSaveAssetClassTargetsMutation() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (targets: Record<string, number>) => {
      const form = new URLSearchParams()
      Object.entries(targets).forEach(([ac, val]) =>
        form.append(`target_${ac.replace(/ /g, '_')}`, String(val))
      )
      return requestForm<{ ok: boolean }>('/api/v1/mf-breakdown/asset-class-targets', form)
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ['mf-breakdown'] }),
  })
}

export function useSaveAllocationTargetsMutation() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (targets: Record<string, number>) => {
      const form = new URLSearchParams()
      Object.entries(targets).forEach(([cat, val]) => form.append(`target_${cat}`, String(val)))
      return requestForm<{ ok: boolean }>('/api/v1/mf-breakdown/allocation-targets', form)
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ['mf-breakdown'] }),
  })
}

export function useSectorList() {
  return useQuery({
    queryKey: breakdownKeys.sectorList,
    queryFn: () => request<string[]>('/api/v1/mf-breakdown/sector-list'),
  })
}

export function useSectorClassifyBatchMutation() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (rows: Array<{ name: string; sector: string }>) =>
      request<SectorClassifyResult>('/api/v1/mf-breakdown/sector-classify-batch', {
        method: 'PATCH',
        body: JSON.stringify(rows),
      }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['mf-breakdown'] }),
  })
}

export function useClassifyBatchMutation() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (rows: Array<{ scheme_isin: string; name: string; category: string }>) => {
      const form = new URLSearchParams()
      rows.forEach((r) => {
        form.append('scheme_isin', r.scheme_isin)
        form.append('name', r.name)
        form.append('category', r.category)
      })
      return requestForm<ClassifyResult>('/api/v1/mf-breakdown/classify-batch', form, 'PATCH')
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ['mf-breakdown'] }),
  })
}
